plugin: add callback specific for validating an updated queue

cmoussa1 opened the pull request:
#### Problem

The multi-factor priority plugin uses the same callback for both _validating_ an updated queue and for _actually updating_ the queue on a user/bank's job. This will be an issue when additional "updateable" attributes are added to the plugin, such as banks.

---

This PR adds a new callback to the plugin, `update_queue_cb ()`, which is responsible for validating an updated queue and will reject the update if it is not valid for the user/bank. It assigns this callback to the `"job.update.attributes.system.queue"` topic string.

As a result of adding the new `update_queue_cb ()` callback, the `job_updated ()` callback is also changed to unpack the "updates" JSON object when unpacking arguments and see if there is an update for the `queue` attribute. It checks its contents and applies its update to the job that requested the queue update.

Lastly, I've added comments at the top of both functions with a short description (mostly so I don't forget how this flow of updating job attributes works in the future 😆).

cmoussa1 committed:
plugin: add update_queue_cb ()

Problem: The multi-factor priority plugin uses the same callback for
validating an updated queue and for actually updating the queue on a
user/bank's job. This will be an issue when additional "updateable"
attributes are added to the plugin, such as banks.

Add a new callback to the plugin, update_queue_cb (), that is
responsible for validating an updated queue and rejects the update
if it is not valid for the user/bank. Assign the callback for the
"job.update.attributes.system.queue" topic string.

As a result of adding the new update_queue_cb () callback, change the
job_updated () callback to unpack the "updates" JSON object when
unpacking arguments and see if there is an update for the queue
attribute. Check its contents for NULL before applying its update to
the job that requested the queue update.

Add a comment at the top of the job_updated () function with a
description of what it does.

cmoussa1 committed:
t1030: update test error message

Problem: The priority plugin prefixes the error message for an invalid
queue with "mf_priority:", which is not included in the "grep" check in
t1030-mf-priority-update-queue.t.

Update the error message to include this prefix.

grondo reviewed (commented):
I found one possible issue inline. I'm actually a little mystified why tests did not fail with this change (it looks like the `bank` can never be detected during a `queue` update). So possibly I'm missing something here...

cmoussa1 commented on src/plugins/mf_priority.cpp:
```diff
@@ -1129,10 +1133,69 @@ static int job_updated (flux_plugin_t *p,
+                                "{s:i, s:s, s?s}",
+                                "userid", &userid,
+                                "value", &queue,
+                                "attributes.system.bank", &bank) < 0)
```
Oops! Sorry about that. Yes, I do mean to be unpacking the `bank` from the jobspec - I think I thought I could shorthand the traditional way of unpacking it from jobspec, just like you suggested above. Since the tests were passing, I figured this was okay too...

Perhaps the reason it passes with this is because in this set of tests, the user is submitting their job under a default bank, and thus `flux_jobtap_jobspec_update_pack ()` is being called to add the bank to the main eventlog?

Either way, I can fix this to unpack it like how I always have.

grondo replied:
Maybe also look into adding a test that exercises this code when the user sets a non-default bank? Just in case :-)

cmoussa1 replied:
sure thing! i can add another set of tests to `t1030` that basically repeats the same workflow, but under a default bank. i'll push up a commit to this branch if that's okay.

cmoussa1 replied:
btw, i was curious, so i went back and tried submitting a job under a non-default bank and using what i had originally proposed in the callback for `job.update.attributes.system.queue`:

```
                                "{s:i, s:s, s?s}",
                                "userid", &userid,
                                "value", &queue,
                                "attributes.system.bank", &bank) < 0)
```

and the tests failed. So thanks for catching that!

cmoussa1 committed:
t1030: add tests that use non-default bank

Problem: t1030-mf-priority-update-queue.t has no tests that test
submitting a pending job under a non-default bank and changing its
queue.

Add some tests that do this.

grondo reviewed (approved):
LGTM!

cmoussa1 commented:
Thanks! Setting MWP here
